lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,354 @@
+			'type'=> 'string',
+			'internaltype'=> 'string',
+			'default'=> '',
+			'required'=> false,
+			'public' => false,
+		),
+		'payment.telephone' => array(
+			'code' => 'payment.telephone',
+			'internalcode'=> 'billingPhone',
+			'label'=> 'Telephone',
+			'type'=> 'string',
+			'internaltype'=> 'string',
+			'default'=> '',
+			'required'=> false,
+			'public' => false,
+		),
+		'payment.email' => array(
+			'code' => 'payment.email',
+			'internalcode'=> 'email',
+			'label'=> 'E-Mail',
+			'type'=> 'string',
+			'internaltype'=> 'string',
+			'default'=> '',
+			'required'=> false,
+			'public' => false,
+		),
+	);
+
+	private $provider;
+	private $basket;
+
+
+	protected function processOrder( \Aimeos\MShop\Order\Item\Iface $order,
+		array $params = [] ) : ?\Aimeos\MShop\Common\Helper\Form\Iface
+	{
+		$parts = \Aimeos\MShop\Order\Item\Base\Base::PARTS_SERVICE
+			| \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT
+			| \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS;
+
+		$base = $this->getOrderBase( $order->getBaseId(), $parts );
+		$data = $this->getData( $base, $order->getId(), $params );
+		$urls = $this->getPaymentUrls();
+		$data['cancelUrl'] = $this->getConfigValue( 'cancelUrl', '' ) ;
+
+		
+		try
+		{
+			$response = $this->sendRequest( $order, $data );
+			$isSuccessful = $response->isSuccessful()  ;
+
+			/*
+			if($isSuccessful){
+				$php_a = $response->getData()['links']['1']['href'] ;
+				echo ($php_a  ) ;
+
+				echo "<script>
+				var approval_ur = '$php_a'; 
+			//    alert(approval_ur);
+				</script>" ;
+
+				echo '
+				<script src="https://www.paypalobjects.com/webstatic/ppplus/ppplus.min.js" type="text/javascript">
+				</script>
+				<div id="ppplus">
+				</div>
+				<script type="application/javascript">
+
+				//  alert(approval_ur) ;
+					var ppp = PAYPAL.apps.PPP({
+					"approvalUrl": approval_ur,
+					"placeholder": "ppplus",
+					"mode": "sandbox",
+					"country": "DE"
+					});
+					setTimeout(function(){ alert("Hello"); }, 3000);
+
+				//	alert(approval_ur) ;
+				</script>
+				
+				';
+			//	die() ;
+			}
+		//	*/
+
+
+			if( $isSuccessful )
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				$this->saveRepayData( $response, $base->getCustomerId() );
+
+				$status = $this->getValue( 'authorize', false ) ? Status::PAY_AUTHORIZED : Status::PAY_RECEIVED;
+				$this->saveOrder( $order->setPaymentStatus( $status ) );
+			}
+			elseif( $response->isRedirect() )
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				return $this->getRedirectForm( $response );
+			}
+			else
+			{
+				$this->saveOrder( $order->setPaymentStatus( Status::PAY_REFUSED ) );
+				throw new \Aimeos\MShop\Service\Exception( $response->getMessage() );
+			}
+		}
+		catch( \Exception $e )
+		{
+			throw new \Aimeos\MShop\Service\Exception( $e->getMessage() );
+		} 
+		$php_a = $response->getData()['links']['1']['href'] ;
```

Comment:
You need to check if it really exists. Also, use speaking variable names.

Reply:
thanks for note ,I have updated my code